commons/src/main/java/org/infinispan/commons/api/functional/Listeners.java

```diff
@@ -0,0 +1,221 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.EntryView.ReadEntryView;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Listener definitions.
+ *
+ * DESIGN RATIONALE:
+ * <ul>
+ *    <li>The current set of listener events that can be fired are related
+ *    to modifications, and deciding between create or modify, or providing
+ *    removed entry information, require reading the previous value. Hence,
+ *    the current set of listeners can only be provided for read-write maps.
+ *    In the future, if for example cache entry visited events are to be
+ *    supported, those would need to be associated with either a read or
+ *    read-write map, and hence the listeners interface would most likely
+ *    be split up.
```

Comment:
Activation, passivation, and expiration events would also have to be available for read-only maps.
Actually, the fact that one object only has a read-only cache view doesn't mean that other objects can't modify the cache, so I'd argue that all listeners should be available for all views.